src/crypto.ts

```diff
@@ -0,0 +1,361 @@
+	if (!provider) {
+		// Load a platform-specific default provider.
+		provider = new Promise(function (resolve, reject) {
+			if (typeof define === 'function' && define.amd) {
+				function loadProvider(mid: string) {
+					require([ mid ], function (_provider: { default: CryptoProvider }) {
+						resolve(_provider);
+					});
+				}
+
+				if (has('host-node')) {
+					loadProvider('./providers/node');
+				}
+				else if (has('webcrypto')) {
+					loadProvider('./providers/webcrypto');
+				}
+				else {
+					loadProvider('./providers/script');
+				}
+			}
+			else if (has('host-node')) {
+				resolve(require('./providers/node'));
+			}
+			else {
+				reject(new Error('Unknown environment or loader'));
+			}
+		});
+	}
+	return provider;
+}
+
+/**
+ * Sets the implementation provider.
+ */
+export function setProvider(_provider: CryptoProvider): void {
+	provider = Promise.resolve(_provider);
+}
+
+/**
+ * An object for hashing a data stream.
+ */
+export interface Hasher<T extends Data> extends Sink<T> {
+	digest: Promise<Binary>;  // read only
+}
+
+/**
+ * A wrapper around a Promise<HashFunction> that will defer calls while a provider is asynchronously loaded.
+ */
+class HasherWrapper<T extends Data> implements Hasher<T> {
+	constructor(hashPromise: Promise<HashFunction>, encoding: string) {
+		Object.defineProperty(this, '_promise', {
+			value: new Promise((resolve, reject) => {
+				hashPromise.then(
+					(hashFunction) => {
+						// When the hash function resolves, create a Hasher and replace this object's methods with
+						// pointers to the corresponding methods on the Hasher.
+						const hasher = hashFunction.create(encoding);
+						this.abort = hasher.abort.bind(hasher);
+						this.close = hasher.close.bind(hasher);
+						this.start = hasher.start.bind(hasher);
+						this.write = hasher.write.bind(hasher);
+						resolve(hasher);
+					},
+					function (error) {
+						reject(error);
+					}
+				);
+			})
+		});
+
+		Object.defineProperty(this, 'digest', {
+			enumerable: true,
+			value: new Promise((resolve, reject) => {
+				this._promise.then(
+					function (hasher) {
+						resolve(hasher.digest);
+					},
+					function (error) {
+						reject(error);
+					}
+				);
+			})
+		});
+	}
+
+	private _promise: Promise<Hasher<T>>;
+
+	digest: Promise<Binary>;
+
+	abort(reason?: Error): Promise<void> {
+		return this._promise.then(function (hasher) {
+			return hasher.abort(reason);
+		});
+	}
+
+	close(): Promise<void> {
+		return this._promise.then(function (hasher) {
+			return hasher.close();
+		});
+	}
+
+	start(error: (error: Error) => void): Promise<void> {
+		return this._promise.then(function (hasher) {
+			return hasher.start(error);
+		});
+	}
+
+	write(chunk: T): Promise<void> {
+		return this._promise.then(function (hasher) {
+			return hasher.write(chunk);
+		});
+	}
+}
+
+/**
+ * A function that can hash a chunk of data.
+ */
+export interface HashFunction {
```

Comment:
I would move this up to under `export interface CryptoProvider`, since it's used in `getHash`.